In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    "Thor eating pizza, Loki is eating pizza, Ironman ate pizza already",
    "Apple is announcing new iphone tomorrow",
    "Tesla is announcing new model-3 tomorrow",
    "Google is announcing new pixel-6 tomorrow",
    "Microsoft is announcing new surface tomorrow",
    "Amazon is announcing new eco-dot tomorrow",
    "I am eating biryani and you are eating grapes"
]

In [12]:
#let's create the vectorizer and fit the corpus and transform them accordingly
v = TfidfVectorizer()
v.fit(corpus)
transform_output = v.transform(corpus)
print(transform_output) # i don't know what this does

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 46 stored elements and shape (7, 28)>
  Coords	Values
  (0, 0)	0.24266547282843012
  (0, 7)	0.24266547282843012
  (0, 10)	0.4028663647756293
  (0, 15)	0.24266547282843012
  (0, 16)	0.11527032701364155
  (0, 17)	0.24266547282843012
  (0, 22)	0.7279964184852904
  (0, 25)	0.24266547282843012
  (1, 4)	0.30652086071532464
  (1, 5)	0.5680354003049032
  (1, 14)	0.5680354003049032
  (1, 16)	0.26982671076064085
  (1, 20)	0.30652086071532464
  (1, 26)	0.30652086071532464
  (2, 4)	0.30652086071532464
  (2, 16)	0.26982671076064085
  (2, 19)	0.5680354003049032
  (2, 20)	0.30652086071532464
  (2, 24)	0.5680354003049032
  (2, 26)	0.30652086071532464
  (3, 4)	0.30652086071532464
  (3, 12)	0.5680354003049032
  (3, 16)	0.26982671076064085
  (3, 20)	0.30652086071532464
  (3, 21)	0.5680354003049032
  (3, 26)	0.30652086071532464
  (4, 4)	0.30652086071532464
  (4, 16)	0.26982671076064085
  (4, 18)	0.5680354003049032
  (4, 20)	0.30652086071532464


In [3]:
#let's print the vocabulary

print(v.vocabulary_)

{'thor': 25, 'eating': 10, 'pizza': 22, 'loki': 17, 'is': 16, 'ironman': 15, 'ate': 7, 'already': 0, 'apple': 5, 'announcing': 4, 'new': 20, 'iphone': 14, 'tomorrow': 26, 'tesla': 24, 'model': 19, 'google': 12, 'pixel': 21, 'microsoft': 18, 'surface': 23, 'amazon': 2, 'eco': 11, 'dot': 9, 'am': 1, 'biryani': 8, 'and': 3, 'you': 27, 'are': 6, 'grapes': 13}


In [10]:
all_feature_names = v.get_feature_names_out()
# all_feature_names is basically a bags of all the words present in our entire dataset in alphabetical order
for word in all_feature_names:
  indx = v.vocabulary_.get(word) # basically each word ko index get gareko xa, although alphabetical order ma thiyo so i don't know sidhai index use gare huntyo enumeratorko
  idf_score = v.idf_[indx] # basically tyo word ko idf kati ho nikaleko ho
  print(f"{word} : {idf_score}")

already : 2.386294361119891
am : 2.386294361119891
amazon : 2.386294361119891
and : 2.386294361119891
announcing : 1.2876820724517808
apple : 2.386294361119891
are : 2.386294361119891
ate : 2.386294361119891
biryani : 2.386294361119891
dot : 2.386294361119891
eating : 1.9808292530117262
eco : 2.386294361119891
google : 2.386294361119891
grapes : 2.386294361119891
iphone : 2.386294361119891
ironman : 2.386294361119891
is : 1.1335313926245225
loki : 2.386294361119891
microsoft : 2.386294361119891
model : 2.386294361119891
new : 1.2876820724517808
pixel : 2.386294361119891
pizza : 2.386294361119891
surface : 2.386294361119891
tesla : 2.386294361119891
thor : 2.386294361119891
tomorrow : 1.2876820724517808
you : 2.386294361119891


In [13]:
print(transform_output.toarray()) # i guess this the the final TF iDF one

[[0.24266547 0.         0.         0.         0.         0.
  0.         0.24266547 0.         0.         0.40286636 0.
  0.         0.         0.         0.24266547 0.11527033 0.24266547
  0.         0.         0.         0.         0.72799642 0.
  0.         0.24266547 0.         0.        ]
 [0.         0.         0.         0.         0.30652086 0.5680354
  0.         0.         0.         0.         0.         0.
  0.         0.         0.5680354  0.         0.26982671 0.
  0.         0.         0.30652086 0.         0.         0.
  0.         0.         0.30652086 0.        ]
 [0.         0.         0.         0.         0.30652086 0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.26982671 0.
  0.         0.5680354  0.30652086 0.         0.         0.
  0.5680354  0.         0.30652086 0.        ]
 [0.         0.         0.         0.         0.30652086 0.
  0.         0.         0.         0.         0.         0.
  0.

In [14]:
import pandas as pd

In [29]:
df = pd.read_csv("./data/ecommerceDataset.csv" , header=None)
df.columns = ['label', 'text']

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50425 entries, 0 to 50424
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   label   50425 non-null  object
 1   text    50424 non-null  object
dtypes: object(2)
memory usage: 788.0+ KB


In [31]:
df['label'].value_counts()

label
Household                 19313
Books                     11820
Electronics               10621
Clothing & Accessories     8671
Name: count, dtype: int64

In [33]:
df['label_num'] = df['label'].map({
    'Household' : 0,
    'Books': 1,
    'Electronics': 2,
    'Clothing & Accessories': 3
})
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50425 entries, 0 to 50424
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   label      50425 non-null  object
 1   text       50424 non-null  object
 2   label_num  50425 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.2+ MB


In [35]:
df.tail(5)

label                                               text  \
50420  Electronics  Strontium MicroSD Class 10 8GB Memory Card (Bl...   
50421  Electronics  CrossBeats Wave Waterproof Bluetooth Wireless ...   
50422  Electronics  Karbonn Titanium Wind W4 (White) Karbonn Titan...   
50423  Electronics  Samsung Guru FM Plus (SM-B110E/D, Black) Colou...   
50424  Electronics                   Micromax Canvas Win W121 (White)   

       label_num  
50420          2  
50421          2  
50422          2  
50423          2  
50424          2

In [59]:
from sklearn.model_selection import train_test_split
import numpy as np
df.dropna(inplace=True)
df.info()
X_train, X_test, y_train, y_test = train_test_split(
    df.text,
    df.label_num,
    test_size=0.2, # 20% samples will go to test dataset
    random_state=2022,
    stratify=df.label_num
) # here the equal distribution is maintained on the basis of hte df.label_num.

# as you can see
print(y_train.value_counts() / len(y_train))
print(y_test.value_counts() / len(y_test))
# you can see they have the same distribution.

<class 'pandas.core.frame.DataFrame'>
Index: 50424 entries, 0 to 50424
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   label      50424 non-null  object
 1   text       50424 non-null  object
 2   label_num  50424 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.5+ MB
label_num
0    0.383004
1    0.234413
2    0.210640
3    0.171943
Name: count, dtype: float64
label_num
0    0.383044
1    0.234408
2    0.210610
3    0.171939
Name: count, dtype: float64


In [61]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report
clf = Pipeline([
     ('vectorizer_tfidf',TfidfVectorizer()), # this is transformer
     ('KNN', KNeighborsClassifier())   # this one is estimator.
])
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.95      0.97      0.96      3863
           1       0.97      0.96      0.96      2364
           2       0.96      0.94      0.95      2124
           3       0.97      0.97      0.97      1734

    accuracy                           0.96     10085
   macro avg       0.96      0.96      0.96     10085
weighted avg       0.96      0.96      0.96     10085



In [62]:
from sklearn.naive_bayes import MultinomialNB
#1. create a pipeline object
clf = Pipeline([
     ('vectorizer_tfidf',TfidfVectorizer()),
     ('Multi NB', MultinomialNB())
])

#2. fit with X_train and y_train
clf.fit(X_train, y_train)


#3. get the predictions for X_test and store it in y_pred
y_pred = clf.predict(X_test)


#4. print the classfication report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.98      0.94      3863
           1       0.98      0.92      0.95      2364
           2       0.97      0.90      0.93      2124
           3       0.98      0.93      0.95      1734

    accuracy                           0.94     10085
   macro avg       0.96      0.93      0.94     10085
weighted avg       0.94      0.94      0.94     10085



In [63]:
from sklearn.ensemble import RandomForestClassifier

#1. create a pipeline object
clf = Pipeline([
     ('vectorizer_tfidf',TfidfVectorizer()),        #using the ngram_range parameter
     ('Random Forest', RandomForestClassifier())
])

#2. fit with X_train and y_train
clf.fit(X_train, y_train)


#3. get the predictions for X_test and store it in y_pred
y_pred = clf.predict(X_test)


#4. print the classfication report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.95      0.98      0.97      3863
           1       0.98      0.98      0.98      2364
           2       0.98      0.94      0.96      2124
           3       0.98      0.97      0.98      1734

    accuracy                           0.97     10085
   macro avg       0.97      0.97      0.97     10085
weighted avg       0.97      0.97      0.97     10085



### since this random forest classifier had the best weighted avg f1-score this is actually preferrred.

In [68]:
import spacy
nlp = spacy.load('en_core_web_sm') # basically we are trying to lemmatize the word and remove stop words and punctuations. so since these factors depend mostly on language like english has different stop words , punctuations and lemmatized variation what language must be specified

# for example

val = nlp('I am having dinner right now! , it is the best dinner i have had in years ')
print(type(val)) # type of spacy token
for token in val:
  if token.is_stop or token.is_punct:
    continue
  print(token.lemma_ , end=" ")


<class 'spacy.tokens.doc.Doc'>
have dinner right good dinner year 

In [70]:

def preprocess(text):
    # remove stop words and lemmatize the text
    doc = nlp(text)
    filtered_tokens = []
    for token in doc:
        if token.is_stop or token.is_punct:
            continue
        filtered_tokens.append(token.lemma_)

    return " ".join(filtered_tokens)
df['preprocessed_txt'] = df['text'].apply(preprocess)

In [71]:
df.head(5)

label                                               text  label_num  \
0  Household  Paper Plane Design Framed Wall Hanging Motivat...          0   
1  Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...          0   
2  Household  SAF 'UV Textured Modern Art Print Framed' Pain...          0   
3  Household  SAF Flower Print Framed Painting (Synthetic, 1...          0   
4  Household  Incredible Gifts India Wooden Happy Birthday U...          0   

                                    preprocessed_txt  
0  Paper Plane Design Framed Wall Hanging Motivat...  
1  SAF Floral Framed Painting Wood 30 inch x 10 i...  
2  SAF UV Textured Modern Art Print Framed Painti...  
3  SAF Flower Print Framed Painting Synthetic 13....  
4  Incredible Gifts India Wooden Happy Birthday U...

In [73]:
X_train , X_test , y_train, y_test  = train_test_split(
    df.preprocessed_txt,
    df.label_num,
    test_size = 0.2,
    random_state=2022,
    stratify=df.label_num
)

In [74]:
#1. create a pipeline object
clf = Pipeline([
     ('vectorizer_tfidf',TfidfVectorizer()),        #using the ngram_range parameter
     ('Random Forest', RandomForestClassifier())
])

#2. fit with X_train and y_train
clf.fit(X_train, y_train)


#3. get the predictions for X_test and store it in y_pred
y_pred = clf.predict(X_test)


#4. print the classfication report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.99      0.97      3863
           1       0.98      0.98      0.98      2364
           2       0.98      0.95      0.96      2124
           3       0.99      0.98      0.98      1734

    accuracy                           0.97     10085
   macro avg       0.98      0.97      0.97     10085
weighted avg       0.97      0.97      0.97     10085



### you can see how removing the stop words, punctuating, and lemmatizing improves the weighted avg of f1-score to 0.97 from 0.96